cmd/commands/flatten_rbd_pvc.go

```diff
@@ -0,0 +1,147 @@
+}
+
+type RBDStatusOutput struct {
+	Watchers []Watcher `json:"watchers"`
+}
+
+var namespace string
+var allowInUse bool
+
+// FlattenRBDPVCCmd represents the rook commands
+var FlattenRBDPVCCmd = &cobra.Command{
+	Use:   "flatten-rbd-pvc",
+	Short: "Flatten the RBD image corresponding to the target RBD PVC",
+	Long: `Flatten the RBD image corresponding to the target RBD PVC.
+The target RBD PVC must be a cloned image and must be created by ceph-csi.
+This command removes the corresponding temporary cloned image[1]
+if the target PVC was cloned from another PVC.
+
+[1]: https://github.com/ceph/ceph-csi/blob/328e4e5a0fa06521c5508b94a5b7ff250ada1681/docs/design/proposals/rbd-snap-clone.md.`,
+	Args: cobra.ExactArgs(1),
+	Run: func(cmd *cobra.Command, args []string) {
+		pvcName := args[0]
+		pvc, err := clientSets.Kube.CoreV1().PersistentVolumeClaims(namespace).Get(cmd.Context(), pvcName, metav1.GetOptions{})
+		if err != nil {
+			logging.Fatal(err, "failed to get pvc %s/%s", namespace, pvcName)
+		}
+		if pvc.DeletionTimestamp != nil {
+			logging.Fatal(fmt.Errorf("pvc %s is deleting", pvcName))
+		}
+		if pvc.Status.Phase != corev1.ClaimBound {
+			logging.Fatal(fmt.Errorf("pvc %s is not bound", pvcName))
+		}
+
+		var shouldDeleteTempImage bool
+		if pvc.Spec.DataSource != nil {
+			switch pvc.Spec.DataSource.Kind {
+			case "PersistentVolumeClaim":
+				shouldDeleteTempImage = true
+			case "VolumeSnapshot":
+			default:
+				logging.Fatal(fmt.Errorf("pvc %s is not a cloned image", pvcName))
+			}
+		}
+
+		pvName := pvc.Spec.VolumeName
+		pv, err := clientSets.Kube.CoreV1().PersistentVolumes().Get(cmd.Context(), pvName, metav1.GetOptions{})
+		if err != nil {
+			logging.Fatal(fmt.Errorf("failed to get pv %s", pvName))
+		}
+		imageName, ok := pv.Spec.CSI.VolumeAttributes["imageName"]
+		if !ok {
+			logging.Fatal(fmt.Errorf("pv %s would not be a CSI PVC", pvName))
+		}
+		poolName, ok := pv.Spec.CSI.VolumeAttributes["pool"]
+		if !ok {
+			logging.Fatal(fmt.Errorf("pv %s would not be a CSI PVC", pvName))
+		}
+
+		if !allowInUse {
+			out, err := exec.RunCommandInOperatorPod(cmd.Context(), clientSets, "rbd", []string{"-p", poolName, "status", imageName, "--format=json"}, operatorNamespace, cephClusterNamespace, false)
+			if err != nil {
+				logging.Fatal(fmt.Errorf("failed to stat %s/%s", poolName, imageName))
+			}
+			var status RBDStatusOutput
+			json.Unmarshal([]byte(out), &status)
+			if len(status.Watchers) >= 1 {
+				logging.Fatal(fmt.Errorf("flatten in-use pvc %s is not allowed. If you want to do, run with `--allow-in-use` option", pvcName))
+			}
+		}
+
+		if shouldDeleteTempImage {
+			deleteTempImage(cmd.Context(), allowInUse, poolName, imageName)
+		}
+		logging.Info("flattening the target RBD image %s/%s", poolName, imageName)
+		_, err = exec.RunCommandInOperatorPod(cmd.Context(), clientSets, "rbd", []string{"-p", poolName, "flatten", imageName}, operatorNamespace, cephClusterNamespace, false)
```

Comment:
flatten takes time depending on the data size, should be make use of `ceph rbd task add flatten`?